### PR TITLE
Feature/minicart spa skin object

### LINKED ISSUE #51: Make Mini Cart Module a Skin Object
Right now, the mini cart is a module and it really should be a skin object that theme developers include, instead of a site administrator managing the module.

For backward compatibility, the mini cart module should still exist for all existing installations.  

New installations should not install the module and only install the skin object.

### PR DESCRIPTION
**Pull request by ismaelTartola**
Merging this branch will introduce the following features:

- Update the MiniCart view to SPA using Jquery. 
- Add MiniCart SkinObject.
- Update the Minicart Module to use the SkinObject.
- Add MiniCartViewModel.
- Add new end points to the CarController in order to provide the interactions needed for the MInicart SPA functionality.
- Mark the CartController Minicart Index Action as Obsolete. 

Closes #51

### CODE REVIEW

File: Website/DesktopModules/Hotcakes/Core/Controllers/CartController.cs [hunk 1]
```diff
@@ -321,6 +321,7 @@ private List<ShippingRateDisplay> GetRates()
         #region Main Cart Actions
 
         // GET: /Cart/
+        [Obsolete("This action is no longer used to load the minicart content.")]
         [NonCacheableResponseFilter]
         public ActionResult Index()
```
Comment (WillStrohl) [inline Website/DesktopModules/Hotcakes/Core/Controllers/CartController.cs:326]: @ismaelTartola Is this the unused endpoint you were telling me about?

File: Website/DesktopModules/Hotcakes/Core/Controllers/CartController.cs [hunk 2]
```diff
@@ -342,6 +343,34 @@ public ActionResult Index()
             return View(model);
         }
 
+        // GET: /MiniCart/
+        [NonCacheableResponseFilter]
+        public ActionResult MiniCart()
```
Comment (WillStrohl) [inline Website/DesktopModules/Hotcakes/Core/Controllers/CartController.cs:348]: @ismaelTartola Should we have any error handling here?  If something goes wrong, nothing is logged, and the UI doesn't know what to do.

File: Website/DesktopModules/Hotcakes/Core/Controllers/CartController.cs [hunk 3]
```diff
@@ -342,6 +343,34 @@ public ActionResult Index()
...
+                TotalQuantity = (CurrentCart != null && CurrentCart.Items != null) ? CurrentCart.Items.Count : 0
+            };
+
+            return View(model);
+        }
+
+        // POST: /MiniCartItems
+        [ActionName("MiniCartItems")]
+        [HccHttpPost]
+        public JsonResult GetMiniCartItems()
```
Comment (WillStrohl) [inline Website/DesktopModules/Hotcakes/Core/Controllers/CartController.cs:361]: @ismaelTartola Should we have any error handling here?  If something goes wrong, nothing is logged, and the UI doesn't know what to do.

File: Website/DesktopModules/Hotcakes/Hotcakes.dnn [hunk 4]
```diff
@@ -1174,6 +1174,47 @@
 			</components>
 		</package>
 
+		<package name="Hotcakes.MiniCartSkinObject" type="SkinObject" version="01.00.00">
```
Comment (WillStrohl) [inline Website/DesktopModules/Hotcakes/Hotcakes.dnn:1177]: @ismaelTartola This version number should always match the version number in the rest of the HCC manifest.

File: Website/DesktopModules/Hotcakes/MiniCart/MiniCartView.ascx [hunk 5]
```diff
@@ -1 +1,5 @@
 ﻿<%@ Control Language="C#" AutoEventWireup="true" CodeBehind="MiniCartView.ascx.cs" Inherits="Hotcakes.Modules.MiniCart.MiniCartView" %>
+<%@ Register TagPrefix="dnn" TagName="MiniCartSpa" Src="../MiniCartSkinObject/MiniCartView.ascx" %>
+
+
+<dnn:MiniCartSpa runat="server" ViewName="<% ViewName %>" id="MiniCartSpa" />
```
Comment (WillStrohl) [inline Website/DesktopModules/Hotcakes/MiniCart/MiniCartView.ascx:5]: @ismaelTartola This appears to be the same control that's being used for the module.  Is that correct?  Can both the module and the skin object be on the same page at the same time?  Does the module still work independently, the same as it did before?

**Review by WillStrohl — CHANGES_REQUESTED** [review]
@ismaelTartola There are a few minor things to address, and a few questions.  :)

File: Website/DesktopModules/Hotcakes/Core/Controllers/CartController.cs (hunk 1 above)
Comment (ismaelTartola) [reply Website/DesktopModules/Hotcakes/Core/Controllers/CartController.cs:326]: Yes. It was used in the past to load the minicart initial information. But after the implementation of the SPA feature there are no call's directed to this action.

File: Website/DesktopModules/Hotcakes/MiniCart/MiniCartView.ascx (hunk 5 above)
Comment (ismaelTartola) [reply Website/DesktopModules/Hotcakes/MiniCart/MiniCartView.ascx:5]: @WillStrohl It's not. What you are noticing here is just a pure naming. I named the main control inside of the skinOject folder just as the main control of the Minicart Module (Of course, they are placed in different folders). I addition to that I like to point that, now the Minicart module itself it's using the skin object, so if we merge this changes every time the Minicart module is used the skin object is used (HCC -> Minicart Module -> Minicart Skin object -> Minicart SPA view). To the question of if the module and the skin object can be used at the same time in a page the answer is YES, no problems there, they will work individually.

File: Website/DesktopModules/Hotcakes/Hotcakes.dnn (hunk 4 above)
Comment (ismaelTartola) [reply Website/DesktopModules/Hotcakes/Hotcakes.dnn]: Changed to 03.07.00.

File: Website/DesktopModules/Hotcakes/Core/Controllers/CartController.cs (hunk 2 above)
Comment (ismaelTartola) [reply Website/DesktopModules/Hotcakes/Core/Controllers/CartController.cs:348]: I tough about that. But when I reviewed the entire controller I did not found any logging system or error handling implemented for any of the actions (Something like a try - catch mechanic). I thought that there was a reason for that. Anyway, In the particular case of MiniCart actions the view will check the view model and based on the status of the view model object will handle the information.

File: Website/DesktopModules/Hotcakes/Core/Controllers/CartController.cs (hunk 1 above)
Comment (WillStrohl) [reply Website/DesktopModules/Hotcakes/Core/Controllers/CartController.cs:326]: @ismaelTartola So, you're saying both the module and the skin object are using the same SPA implementation now?

File: Website/DesktopModules/Hotcakes/Core/Controllers/CartController.cs (hunk 2 above)
Comment (WillStrohl) [reply Website/DesktopModules/Hotcakes/Core/Controllers/CartController.cs:348]: @ismaelTartola In the very least, there should be some sort of `try ... catch` in the endpoint itself.  If something goes wrong, the error is logged to the Log4NET log file.  There should also be some sort of generic error message in the UI.  Something stating, "An unknown error has occurred. If this continues to happen, please contact the store directly."

File: Website/DesktopModules/Hotcakes/Core/Controllers/CartController.cs (hunk 3 above)
Comment (WillStrohl) [reply Website/DesktopModules/Hotcakes/Core/Controllers/CartController.cs:361]: @ismaelTartola In the very least, there should be some sort of `try ... catch` in the endpoint itself.  If something goes wrong, the error is logged to the Log4NET log file.  There should also be some sort of generic error message in the UI.  Something stating, "An unknown error has occurred. If this continues to happen, please contact the store directly."  

You might want to use the built-in UI messaging.  https://uxguide.dnnsoftware.com/UIPatterns/ModuleMessages.html

File: Website/DesktopModules/Hotcakes/MiniCart/MiniCartView.ascx (hunk 5 above)
Comment (WillStrohl) [reply Website/DesktopModules/Hotcakes/MiniCart/MiniCartView.ascx:5]: @ismaelTartola Okay... Just so I'm sure I understand, both the SPA skin object and the module are using the same underlying code, but the UI is initiated from different UI/UX locations.   Is that what you're saying?

File: Website/DesktopModules/Hotcakes/Core/Controllers/CartController.cs (hunk 1 above)
Comment (ismaelTartola) [reply Website/DesktopModules/Hotcakes/Core/Controllers/CartController.cs:326]: Yes.

File: Website/DesktopModules/Hotcakes/Core/Controllers/CartController.cs (hunk 2 above)
Comment (ismaelTartola) [reply Website/DesktopModules/Hotcakes/Core/Controllers/CartController.cs:348]: That's right. I think that to. I will add it. I just haven't due to that there is no error handling in any of the rest of the controller actions. I thought that a very specific reason exit for that and I follow the same code style.

Comment (WillStrohl) [reply Website/DesktopModules/Hotcakes/Core/Controllers/CartController.cs:348]: @ismaelTartola  I look forward to your follow-up comment after our conversation during today's meeting

Comment (ismaelTartola) [reply Website/DesktopModules/Hotcakes/Core/Controllers/CartController.cs:351]: @WillStrohl Like I commented in slack today I haven't been able to find any error handling or logging feature implementation, so I have added a custom implementation of the features(logging and error handling) to the actions involved in the minicart skin object functionalities. But that is so rare, I have even take the liberty to explore the many of the other hotcackes modules controllers and I'm haven't found a single implementation for logging or error handling anywhere. That concern me, I have to point that I really think we are missing something. I my mind I just can't think that this the way this solution is implemented.

File: Website/DesktopModules/Hotcakes/Core/Controllers/CartController.cs (hunk 3 above)
Comment (ismaelTartola) [reply Website/DesktopModules/Hotcakes/Core/Controllers/CartController.cs:371]: @WillStrohl Like I commented in slack today I haven't been able to find any error handling or logging feature implementation, so I have added a custom implementation of the features(logging and error handling) to the actions involved in the minicart skin object functionalities. But that is so rare, I have even take the liberty to explore the many of the other hotcackes modules controllers and I'm haven't found a single implementation for logging or error handling anywhere. That concern me, I have to point that I really think we are missing something. I my mind I just can't think that this the way this solution is implemented.

**Review by WillStrohl — APPROVED** [review]
Looks good!

**Pull request merged**